File: diagram/ProbabilisticOperator.py
"""
Author: Francesco Nieri
Class representing a probabilistic operator in an outcome diagram
"""

from diagram.ObservationPoint import ObservationPoint
from diagram.DiagramComponent import DiagramComponent
from utils.EmpiricalCDF import EmpiricalCDF


class ProbabilisticOperator(DiagramComponent):
    def __init__(self, name, following_components):
        super().__init__(name)
        self.following_components_and_probabilities: dict[ObservationPoint, int] = following_components
        if not self.check_probabilities_sum():
            raise ValueError("The sum of probabilities must equal 1.")

    def check_probabilities_sum(self) -> bool:
        """
        Check if the probabilities supplied actually sum to 1
        :return:
        """
        total_prob = sum(self.following_components_and_probabilities.values())
        return abs(total_prob - 1) < 1e-6

    def calculate_dq(self, *args) -> EmpiricalCDF:
        """
        Calculate the DeltaQ for the probabilistic operator
        :param args:
        :return:
        """
        cdfs = []
        for component, probability in self.following_components_and_probabilities.items():
            cdf = component.calculate_dq() * probability
            cdfs.append(cdf)
        cdf = cdfs[0].add_cdfs(cdfs[1:])
        return cdf

    def is_plottable(self):
        return False

File: src/c++/diagram/ProbabilisticOperator.cpp
/**
 * @author Francesco Nieri
 * @date 25/10/2024
 * Class representing a probabilstic operator in a DeltaQ system
 */


#include "ProbabilisticOperator.h"

#include <utility>
#include "../maths/DeltaQOperations.h"

ProbabilisticOperator::ProbabilisticOperator(const std::string& name, std::map<std::shared_ptr<DiagramComponent>, double> followingComponents) :
    DiagramComponent{name},
    followingComponentAndProbabilities{std::move(followingComponents)}
    {}  


DeltaQ ProbabilisticOperator::calculateDeltaQ(const System& system, const DeltaQ& deltaQ) {
    std::vector<double> resultingCdf;
    std::vector<DeltaQ> scaledDeltaQs;
    
     for (const auto& entry : followingComponentAndProbabilities) {
        const std::shared_ptr<DiagramComponent> component = entry.first;
        const double probability = entry.second;

        DeltaQ currentDeltaQ = component->calculateDeltaQ(system, deltaQ);

        scaledDeltaQs.push_back(currentDeltaQ * probability);
    }

    DeltaQ result = scaledDeltaQs[0];
    for (std::size_t i = 1; i < scaledDeltaQs.size(); ++i) {
        result = result + scaledDeltaQs[i];
    }
    return result;
} 